### PR TITLE
Multi angle detection

### PR DESCRIPTION
**Pull request by harisreedhar**

### CODE REVIEW

File: facefusion/face_analyser.py [hunk 1]
```diff
@@ -7,13 +7,13 @@
...
+from facefusion.face_helper import estimate_matrix_by_face_landmark_5, warp_face_by_face_landmark_5, warp_face_by_translation, create_static_anchors, distance_to_face_landmark_5, distance_to_bounding_box, convert_to_face_landmark_5, normalize_bounding_box, convert_to_rotated_bounding_box, create_rotation_matrix_with_size, transform_bounding_box, transform_points
 from facefusion.face_store import get_static_faces, set_static_faces
 from facefusion.execution import apply_execution_provider_options
 from facefusion.download import conditional_download
 from facefusion.filesystem import resolve_relative_path, is_file
 from facefusion.thread_helper import thread_lock, thread_semaphore, conditional_thread_semaphore
-from facefusion.typing import VisionFrame, Face, ModelSet, BoundingBox, FaceLandmarkSet, FaceLandmark5, FaceLandmark68, Score, FaceScoreSet, Embedding
+from facefusion.typing import VisionFrame, Face, ModelSet, BoundingBox, FaceLandmarkSet, FaceLandmark5, FaceLandmark68, Score, FaceScoreSet, Embedding, RotatedBoundingBox
```
Comment (henryruhs) [inline facefusion/face_analyser.py:16]: cosmetic but change order BoundingBox, RotatedBoundingBox

File: facefusion/face_analyser.py [hunk 2]
```diff
@@ -534,9 +543,23 @@ def get_many_faces(vision_frames : List[VisionFrame]) -> List[Face]:
...
+				rotated_bounding_boxes_all = []
+				face_landmarks_5_all = []
+				face_scores_all = []
+
+				for angle in facefusion.globals.face_detector_angles:
+					if angle == 0:
+						bounding_boxes, face_landmarks_5, face_scores = detect_faces(vision_frame)
+					else:
+						bounding_boxes, face_landmarks_5, face_scores = detect_faces_with_rotation(vision_frame, angle)
+					rotated_bounding_boxes = [ convert_to_rotated_bounding_box(bounding_box, angle) for bounding_box in bounding_boxes ]
```
Comment (henryruhs) [inline facefusion/face_analyser.py:556]: I do not fully understand why convert_to_rotated_bounding_box() is needed as you pass rotated_bounding_boxes already to the method?

File: facefusion/face_helper.py [hunk 3]
```diff
@@ -115,6 +123,28 @@ def normalize_bounding_box(bounding_box : BoundingBox) -> BoundingBox:
...
 
 
+def transform_points(points : numpy.ndarray[Any, Any], matrix : Matrix) -> numpy.ndarray[Any, Any]:
+	points = points.reshape(-1, 1, 2)
+	points = cv2.transform(points, matrix) # type:ignore[assignment]
+	points = points.reshape(-1, 2)
+	return points
+
+
+def transform_bounding_box(bounding_box : BoundingBox, matrix : Matrix) -> BoundingBox:
```
Comment (henryruhs) [inline facefusion/face_helper.py:133]: prepare bounding box for xxx? the method does not say too much what it all does

File: facefusion/face_helper.py [hunk 4]
```diff
@@ -101,6 +101,14 @@ def create_bounding_box_from_face_landmark_68(face_landmark_68 : FaceLandmark68)
 	return bounding_box
 
 
+def create_rotation_matrix_with_size(angle : float, size : Size) -> Tuple[Matrix, Size]:
```
Comment (henryruhs) [inline facefusion/face_helper.py:104]: lets keep same naming rotated matrix similar to rotated bounding box

File: facefusion/face_analyser.py [hunk 5]
```diff
@@ -534,9 +543,23 @@ def get_many_faces(vision_frames : List[VisionFrame]) -> List[Face]:
 			if static_faces:
 				faces = static_faces
 			else:
-				bounding_boxes, face_landmarks_5, face_scores = detect_faces(vision_frame)
-				if bounding_boxes and face_landmarks_5 and face_scores and facefusion.globals.face_detector_score > 0:
-					faces = create_faces(vision_frame, bounding_boxes, face_landmarks_5, face_scores)
+				bounding_boxes_all = []
```
Comment (henryruhs) [inline facefusion/face_analyser.py:546]: those temporary _all variables are not needed as you extend them once!?

File: facefusion/face_analyser.py [hunk 6]
```diff
@@ -362,6 +362,16 @@ def detect_faces(vision_frame: VisionFrame) -> Tuple[List[BoundingBox], List[Fac
 	return bounding_boxes, face_landmarks_5, face_scores
 
 
+def detect_faces_with_rotation(vision_frame : VisionFrame, angle : float) -> Tuple[List[BoundingBox], List[FaceLandmark5], List[Score]]:
+	rotation_matrix, rotated_size = create_rotation_matrix_with_size(angle, vision_frame.shape[:2][::-1])
+	rotated_vision_frame = cv2.warpAffine(vision_frame, rotation_matrix, rotated_size)
+	inverse_rotation_matrix = cv2.invertAffineTransform(rotation_matrix)
+	bounding_boxes, face_landmarks_5, face_scores = detect_faces(rotated_vision_frame)
+	bounding_boxes = [ transform_bounding_box(bounding_box, inverse_rotation_matrix) for bounding_box in bounding_boxes ]
```
Comment (henryruhs) [inline facefusion/face_analyser.py:370]: I think this transform methods could be renamed to rotate_back

File: facefusion/globals.py [hunk 7]
```diff
@@ -25,6 +25,7 @@
 face_detector_model : Optional[FaceDetectorModel] = None
 face_detector_size : Optional[str] = None
 face_detector_score : Optional[float] = None
+face_detector_angles : Optional[List[float]] = None
```
Comment (henryruhs) [inline facefusion/globals.py:28]: introduce a FaceDetectorAngle literal please

FaceDetectorAngles = Literal[0, 90, 180, 270]

File: facefusion/core.py [hunk 8]
```diff
@@ -85,8 +85,9 @@ def create_program() -> ArgumentParser:
 	group_face_analyser.add_argument('--face-detector-model', help = wording.get('help.face_detector_model'), default = config.get_str_value('face_analyser.face_detector_model', 'yoloface'), choices = facefusion.choices.face_detector_set.keys())
 	group_face_analyser.add_argument('--face-detector-size', help = wording.get('help.face_detector_size'), default = config.get_str_value('face_analyser.face_detector_size', '640x640'), choices = suggest_face_detector_choices(program))
 	group_face_analyser.add_argument('--face-detector-score', help = wording.get('help.face_detector_score'), type = float, default = config.get_float_value('face_analyser.face_detector_score', '0.5'), choices = facefusion.choices.face_detector_score_range, metavar = create_metavar(facefusion.choices.face_detector_score_range))
+	group_face_analyser.add_argument('--face-detector-angles', help = wording.get('help.face_detector_angles'), type = float, default = config.get_float_list('face_analyser.face_detector_angles', '0.0'), choices = facefusion.choices.face_detector_angles, nargs = '+')
```
Comment (henryruhs) [inline facefusion/core.py:88]: Did you miss adding the choices? facefusion.choices.face_detector_angles

File: facefusion/face_helper.py [hunk 9]
```diff
@@ -115,6 +123,28 @@ def normalize_bounding_box(bounding_box : BoundingBox) -> BoundingBox:
...
+
+def transform_bounding_box(bounding_box : BoundingBox, matrix : Matrix) -> BoundingBox:
+	points = numpy.array(
+	[
+		[ bounding_box[0], bounding_box[1] ],
+		[ bounding_box[2], bounding_box[1] ],
+		[ bounding_box[2], bounding_box[3] ],
+		[ bounding_box[0], bounding_box[3] ]
+	]
+	).astype(numpy.float32)
```
[the remark this answers is not in the document]
Comment (harisreedhar) [reply facefusion/face_helper.py:141]: by default numpy array is float64

File: facefusion/face_helper.py (hunk 3 above)
Comment (harisreedhar) [reply facefusion/face_helper.py:133]: with `matrix` bounding box can be scaled, rotated or translated. thats why name transform is given.

File: facefusion/face_helper.py [hunk 10]
```diff
@@ -115,6 +123,28 @@ def normalize_bounding_box(bounding_box : BoundingBox) -> BoundingBox:
 	return numpy.array([ x1, y1, x2, y2 ])
 
 
+def transform_points(points : numpy.ndarray[Any, Any], matrix : Matrix) -> numpy.ndarray[Any, Any]:
```
[the remark this answers is not in the document]
Comment (harisreedhar) [reply facefusion/face_helper.py:126]: same like bounding_box

**Pull request merged**